packages/protocol/contracts/layer1/preconf/impl/Protector.sol

```diff
@@ -0,0 +1,147 @@
+
+    /// @inheritdoc IProtector
+    function addSigner(address _signer, bytes[] memory _signatures) external {
+        require(!signers[_signer], SignerAlreadyExists());
+
+        bytes32 digest = keccak256(
+            abi.encode(LibPreconfConstants.ADD_PROTECTOR_SIGNER_DOMAIN_SEPARATOR, nonce, _signer)
+        );
+        _verifySignatures(digest, _signatures);
+
+        signers[_signer] = true;
+        unchecked {
+            ++numSigners;
+            ++nonce;
+        }
+
+        emit SignerAdded(_signer);
+    }
+
+    /// @inheritdoc IProtector
+    function removeSigner(address _signer, bytes[] memory _signatures) external {
+        require(signers[_signer], SignerDoesNotExist());
+        require(numSigners > signingThreshold, CannotRemoveSignerWhenThresholdIsReached());
+
+        bytes32 digest = keccak256(
+            abi.encode(LibPreconfConstants.REMOVE_PROTECTOR_SIGNER_DOMAIN_SEPARATOR, nonce, _signer)
+        );
+        _verifySignatures(digest, _signatures);
+
+        delete signers[_signer];
+        unchecked {
+            --numSigners;
+            ++nonce;
+        }
+
+        emit SignerRemoved(_signer);
+    }
+
+    /// @inheritdoc IProtector
+    function updateSigningThreshold(
+        uint128 _signingThreshold,
+        bytes[] memory _signatures
+    )
+        external
+    {
+        require(_signingThreshold <= numSigners, InvalidSigningThreshold());
+
+        bytes32 digest = keccak256(
+            abi.encode(
+                LibPreconfConstants.UPDATE_PROTECTOR_SIGNING_THRESHOLD_DOMAIN_SEPARATOR,
+                nonce,
+                _signingThreshold
+            )
+        );
+        _verifySignatures(digest, _signatures);
+
+        signingThreshold = _signingThreshold;
+        unchecked {
+            ++nonce;
+        }
+
+        emit SigningThresholdUpdated(_signingThreshold);
+    }
+
+    // Internal functions
+    // -----------------------------------------------------------------------------------
+
+    function __Protector_init(uint128 _signingThreshold, address[] memory _signers) internal {
+        require(_signingThreshold <= _signers.length, InvalidSigningThreshold());
+
+        for (uint256 i; i < _signers.length; ++i) {
+            signers[_signers[i]] = true;
+        }
+
+        numSigners = uint128(_signers.length);
+        signingThreshold = _signingThreshold;
+    }
+
+    function _verifySignatures(bytes32 digest, bytes[] memory _signatures) internal view {
+        address lastSigner;
+        address currentSigner;
+
+        for (uint256 i; i < _signatures.length; ++i) {
+            // Recover the signer from the signature
+            currentSigner = ECDSA.recover(digest, _signatures[i]);
+            require(signers[currentSigner], NotAnExistingSigner());
+
+            // To prevent replay of signatures
+            require(lastSigner < currentSigner, SignersMustBeSortedInAscendingOrder());
```

Comment:
Why is the order important of `_signatures` important?

Reply:
Ah, I see. But don't we also need a check that the length of `_signatures` is same as `numSigners`?

Reply:
Ah, @mikhailUshakoff you already had similar comment above.